hal-core/src/mem/page/alloc.rs

```diff
@@ -0,0 +1,571 @@
+    fn alloc_range(&self, size: S, len: usize) -> Result<PageRange<PAddr, S>> {
+        let span = tracing::trace_span!("alloc_range", size = size.as_usize(), len);
+        let _e = span.enter();
+
+        if size.as_usize() > self.min_size {
+            // TODO(eliza): huge pages should work!
+            tracing::error!(
+                requested.size = size.as_usize(),
+                requested.len = len,
+                "cannot allocate; huge pages are not currently supported!"
+            );
+            return Err(AllocErr::oom());
+        }
+
+        // This is the minimum order necessary for the requested allocation ---
+        // the first free list we'll check.
+        let order = self.order_for(size, len)?;
+        tracing::trace!(?order);
+
+        // Try each free list, starting at the minimum necessary order.
+        for (idx, free_list) in self.free_lists.as_ref()[order..].iter().enumerate() {
+            tracing::trace!(curr_order = idx + order);
+
+            // Is there an available block on this free list?
+            let mut free_list = free_list.lock();
+            if let Some(mut block) = free_list.pop_back() {
+                let block = unsafe { block.as_mut() };
+                tracing::trace!(?block, ?free_list, "found");
+
+                // Unless this is the free list on which we'd expect to find a
+                // block of the requested size (the first free list we checked),
+                // the block is larger than the requested allocation. In that
+                // case, we'll need to split it down and push the remainder onto
+                // the appropriate free lists.
+                if idx > 0 {
+                    let curr_order = idx + order;
+                    tracing::trace!(?curr_order, ?order, "split down");
+                    self.split_down(block, curr_order, order);
+                }
+
+                // Change the block's magic to indicate that it is allocated.
+                // /!\ EXTREMELY SERIOUS WARNING /!\
+                // This is ACTUALLY LOAD BEARING. If a block is allocated and
+                // the first word isn't immediately written to, it *might* still
+                // appear to be "free" when its buddy is deallocated, and the
+                // buddy may want to merge with it. When the buddy tries to
+                // remove the block from the free list, though, it will be
+                // surprised to find that the block is not, in fact, free!
+                //
+                // Therefore, we need to clobber the block magic, even if it
+                // will *probably* be immediately overwritten by whoever
+                // allocated the block --- they *might* not overwrite it, or
+                // someone else might free the buddy of this block before the
+                // user of this block writes to it!
+                block.make_busy();
```

Comment:
This is scary that we're depending on the user not using the free magic in their heap-allocated payload D:

Reply:
So, technically, the free list is the authoritative source for whether or not a block is free, the magic is just a hint so that we don't have to traverse the whole free list when a block is definitely busy. However, we currently panic if the magic indicates that the block is free, but it's not on the free list...which is incorrect way to handle a false positive. 

I changed it in 14ec7b9dc343ed728d9d74081dd2488f61e36a37 so that the free magic is *strictly* an optimization hint, rather than relying on it for correctness.